File: go/various/maxsubarray.go
// Implements the maximum subarray algorithm on a slice of ints
package various

func maxCrossingSubarray(array []int, from, to int) (li, ri, sum int) {
	mid := (from + to) / 2
	// left index, left side's sum, new running sum
	li, lsum, nsum := mid-1, array[mid-1], array[mid-1]
	for n := li - 1; n >= from; n-- {
		nsum += array[n]
		if nsum > lsum {
			lsum = nsum
			li = n
		}
	}
	ri, rsum, nsum := mid, array[mid], array[mid]
	for n := ri + 1; n < to; n++ {
		nsum += array[n]
		if nsum > rsum {
			rsum = nsum
			ri = n
		}
	}
	return li, ri + 1, lsum + rsum // one after last valid index
}

func MaxSubarray(array []int, from, to int) (li, ri, sum int) {
	if from >= to-1 {
		if to-from == 0 {
			return from, to, 0
		}
		return from, to, array[from]
	} else {
		lli, lri, lv := MaxSubarray(array, from, (from+to)/2)
		rli, rri, rv := MaxSubarray(array, (from+to)/2, to)
		cli, cri, cv := maxCrossingSubarray(array, from, to)
		if lv > rv && lv > cv {
			return lli, lri, lv // left's left index, right index, sum
		} else if rv > lv && rv > cv {
			return rli, rri, rv // right's left index, right index, sum
		} else {
			return cli, cri, cv // crossing left index, right index, sum
		}
	}
}

File: c/various/max_subarray.c
#include "max_subarray.h"

// maximum subarray from left to right
max_info max_subarray(int *arr, int l, int r) {
  max_info max_now = {l, l, 0};
  if (r - l <= 1) {
    max_now.r = r;
    if (r == l) {
      return max_now;
    }
    max_now.sum = arr[l];
    return max_now;
  }
  max_now.sum = arr[l];
  max_info max_so_far = {l, l, arr[l]}; 
  for (l += 1; l < r; l++) {
    if (max_now.sum + arr[l] > arr[l]) { // yet net higher than start
      max_now.r = l;
      max_now.sum += arr[l];
    } else { // new lowest low
      max_now.l = l;
      max_now.r = l;
      max_now.sum = arr[l];
    }
    if (max_now.sum > max_so_far.sum) {
      max_so_far = max_now;
    }
  }
  max_so_far.r++; // increment to one past the actual right index
  return max_so_far;
}

